R/CDPSession.R

```diff
@@ -317,18 +421,17 @@ CDPConnexion <- R6::R6Class(
 )
 
 chr_get_ws_addr <- function(host = "localhost", port = 9222, secure = FALSE) {
-  url <- build_url(host, port, secure)
   "!DEBUG Retrieving Chrome websocket entrypoint at http://localhost:`port`/json ..."
-  open_debuggers <- tryCatch(
-    jsonlite::read_json(paste0(url, "/json"), simplifyVector = TRUE),
-    error = function(e) list()
-  )
-
-  address <- open_debuggers$webSocketDebuggerUrl[open_debuggers$type == "page"]
+  targets <- list_targets(host, port, secure)
+  active_target <- purrr::detect(targets, ~ identical(.x$type, "page"))
+  address <- active_target$webSocketDebuggerUrl
   if (is.null(address))
     "!DEBUG ...websocket entrypoint unavailable."
   else
     "!DEBUG ...found websocket entrypoint `address`"
 
+  if(isTRUE(secure)) {
+    address <- httr::modify_url(address, scheme = "wss")
+  }
   address # NULL when fails
```

Comment:
Great test! Thanks!

> And chrome is still alive, I think either it shouldn't be or I should be able to reconnect ?
Chrome is alive without any opened tab (there is one default tab when Chrome opens). So, there is no `page` endpoint. 

There are many possible turnarounds. IMO, we could test whether Chrome has an opened tab: if there is no tab, `chrome$connect()` could create a new tab first.

Reply:
Through https://github.com/RLesur/crrri/pull/56/commits/0ec3a3fc83226158377b1e02aa0382abe93a8863, the `connect()` method of `CDPSession` is now a private method: since the websocket connection is automatically connected, there is no need to expose it as a public member.
In order to resolve the problem due to missing tab, https://github.com/RLesur/crrri/pull/56/commits/6626cbc5a61dd7de035415348ca6e30654e4b4c5 creates a new `connectToNewTab()` method in `CDPRemote`. The `connect()` method with `.target_id = "default"` now checks whether there is one available target. If there is no target, a new one is created using `connectToNewTab()`.